guides/routing/concepts/README.md

```diff
@@ -0,0 +1,103 @@
+
+```c#
+Response Run(RoutingInfo routingInfo, Func<Response> next)
+{
+  Logger.LogRequestStart(routingInfo);
+  Response originalResponse = next();
+  Logger.LogRequestEnd(routingInfo);
+  return originalResponse;
+}
+```
+
+If you want to prevent the original handling from happening, just don't call next and return an alternative response:
+
+```c#
+Response Run(RoutingInfo routingInfo, Func<Response> next)
+{
+  if(IsAuthorized(routingInfo)))
+  {
+    return next();
+  }
+  else
+  {
+    return Response.FromStatusCode(403);
+  }
+}
+```
+
+Middleware usually decide weather (and how) to act on the request basing on the `RoutingInfo.SelectedPageType` and its associated attributes.
+
+### Context
+
+Context is the data represented in URL arguments. Page `PersonPage` associated with URL `/people/person/{id}` would probably have Context of type `Person` (and would probably be `IBound<Person>`). If the type of the Page is `IBound<T>`, the Context type will be inferred to `T`. You can also explicitly specify the Context type by implementing the `IPageContext<T>` interface.
+
+Context was introduced to let the middleware reason about the object(s) associated with the Page before the Page is constructed. If the Page is `IBound<T>` we can set its `Data` property to T and then act on it (e.g. decide weather the user is authorized to interact with the Page). But before the page is constructed (when majority of Middleware is run) the Context can't be retrieved from the Page object, since there is no Page object yet. That's why the `Context` property of the `RoutingInfo` has been introduced.
+
+It's usually the best to register the `ContextMiddleware`, since it will automatically set the `RoutingInfo.Context` property if the Context type is a database type (which is almost always true). If you want to set the Context your way, just create and register your own Middleware that does it.
+
+If you want to customise the the Context type, way that is created or handled for your page, follow this example:
+
+```c#
+[Url("/items/{0}/{1}")]
+public partial class ItemPage: IPageContext<Tuple<Item, SubItem>>, IBound<SubItem>
+//                             ^^-- this way you override the Context type, which would be inferred to SubItem
+{
+    private Item _item;
+    
+    [UriToContext]
+    // this attribute marks the method that will be used to create the Context
+    public static Tuple<Item, SubItem> UriToContext(string[] args)
+    // its return type matches the Context type. It accepts arguments from URL
+    {
+        return Tuple.Create(Items.ById(args[0]), SubItems.ById(args[1]));
+    }
+    
+    public void HandleContext(Tuple<Item, SubItem> context)
+    // this method is defined in the IPageContext interface. It is invoked after the Page is created
+    {
+        this._item = context.Item;
+        this.Data = context.SubItem;
+    }
+}
+```
```

Comment:
I think that it would be nice to give an example of custom context together with permission check. This is like "full scope" usage.